modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+  while operating in that it enables us to manage 'best-case', 'average-case',
+  and 'worst-case' scenarios by making multiple passes over each segment. In the
+  case where there is no oversubscription, the best-case will always be achieved
+  (considering any other conditions are also met), while in the case of oversubscription
+  or, for example a near empty bag, we fall into the 'average-case', etc. Examples
+  of 'best-case' scenarios for a removal would be when a segment is unlocked
+  and contains some elements we can drain, and the 'average-case' would be to find
+  any segment, unlocked or not, that contains elements we can drain, and so on.
+
+  This data structure also employs its own load balancing algorithm, beginning
+  with a best-effort round-robin algorithm such that each task begins searching
+  at another segment (with the added benefit of reducing overall contention), in
+  particular is useful for locally distributing insertion operations. Next it also
+  employs a work-stealing algorithm that horizontally steals a ratio of elements
+  across segments. A segment is considered 'horizontal' if it shares the same
+  segment index within [0, here.maxTaskPar); those other segments likely are also
+  benefitted from local distribution from the round-robin algorithm discussed above.
+  We also steal a ratio of elements, say 25%, because it leaves all other segments
+  with 75% of their work; the more elements they have, the more we take, but the
+  less they have, the less we steal; this also has the added benefit of reducing
+  unnecessary work stealing between segments when the bag is nearly emptied. Furthermore,
+  stealing horizontally ensures that all segments remain filled and that we still
+  achieve parallelism across segments for removal operations. Lastly, we attempt
+  to steal a maximum of `N / sizeof(eltType)`, where N is some size in megabytes
+  (representing how much data can be sent in one network request), which keeps
+  down excessive communication.
+
+  This data structure does not come without flaws; as work stealing is dynamic
+  and triggered on demand, work stealing can still be performed in excess, which
+  dramatically causes a performance drop. Furthermore, a severe imbalance across
+  nodes, such as an unfair distribution of work to a single or small subset of nodes,
+  may also causes an equally severe performance degradation. To circumvent this,
+  we allow the user to determine manually when to 'balance', which will evenly distribute all
+  elements across nodes. This data structure scales in terms of nodes, processors per node,
+  and even work load. The larger the work load, the more data that gets stolen when
+  stealing work, and better locality of elements among segments. As well, to
+  achieve true parallelism, usage of a privatized instance is a requirement, as
+  it avoids the overhead of remotely accessing class fields, bounding performance
+  on communication.

```

Comment:
@mppf 

Here are some potential issues I would need to know how to handle with record wrapping...

1) How can the user use these parallel-safe data structures in a parallel context, such as `forall` and `coforall`? Engin informed me of `pragma "default intent is ref if modified"` which does seem to work, but just to be sure.

2) How do we manage copies of the original record wrapped collection if the user were to make a copy by-value? For example, if you were to allocate the structure in a small scope but have it be used outside of it, and the original goes out of scope, then how do we ensure we do not clean this up? Engin suggested [`SharedObject`](http://chapel.cray.com/docs/latest/modules/packages/SharedObject.html?module-SharedObject), and I could overload assignment operator to automatically update the reference to it. Just to make sure you'd want this change.

3) The record-wrapped types are not generic on the implementation, only on the element itself. The type cannot be used polymorphically/dynamically at all. `CollectionImpl` becomes more of a 'pseudo-interface', a simple 'promise' to implement at least those methods (although at this point, do we really need the base `Collection` anymore? It'd be less restrictive to just remove it all-together and not use any inheritance.) This is a rather drastic change, I'm not sure how long this would take me.


Finally, would this need to be done by the initial PR and not in a separate one? If we're going to be making this change, then `Collection` kind of goes away entirely, or if it does exist it wouldn't really be entirely useful.